### ISSUE TITLE
index_select, take can't receive zero dim index

### ISSUE THREAD

**Issue by mttbx** [pytorch/pytorch#6919]
Hi,
It seems that some torch api function like torch.take and torch.index_select can't make it right when the index is 0 dims.

for example:
 I need pick the some element from a vector.
```
a = torch.IntTensor([1, 2, 3])

// implement 1.
index1 = torch.nonzero(a>3)[:, 0]
index2 = torch.nonzero(a>1)[:, 0]
a1 = torch.index_select(a, 0, index1) // error! the index dim is zero
a2 = torch.index_select(a, 0, index2)
c = torch.cat((a1, a2))

// implement 2.
index1 = torch.nonzero(a>3)[:, 0]
index2 = torch.nonzero(a>1)[:, 0]
a1 = torch.take(a, index1) // error! the index dim is zero
a2 = torch.take(a, index2)
c = torch.cat((a1, a2))

// implement 3.
a1 = torch.masked_select(a, a>3)
a2 = torch.masked_select(a, a>1)
c = torch.cat((a1, a2))
```

implement 1 and 2 can't work. I can only use implement 3. However torch.masked_select can't be directly used for non-vector situation. So I need some work around, that's not good...
Can we make torch.index_select and torch.take do receive  zero dim index and return zero dim tensor?
Besides, torch.cat can do the job when some of the operand is zero dim...

**Comment by fmassa** [comment]
I can't reproduce in v0.4.

Also, all the cases you have presented can be more simply implemented using standard advanced indexing.

```python
a = torch.IntTensor([1, 2, 3])
a1 = a[a > 0]

# or for byte masks
a = torch.arange(9).reshape(3, 3)
index = torch.tensor([False, True, False])
b = a[index]
c = a[:, index]
```

**Comment by gchanan** [comment]
closing due to @fmassa's comment, please let me know if you think this is incorrect.
